plugins/preview/plugin.js

```diff
@@ -8,128 +8,262 @@
-		init: function( editor ) {
+	function createPreviewHtml( editor ) {
+		var pluginPath = CKEDITOR.plugins.getPath( 'preview' ),
+			config = editor.config,
+			title = editor.lang.preview.preview,
+			baseTag = generateBaseTag();
 
-			// Preview is not used for the inline creator.
-			if ( editor.elementMode == CKEDITOR.ELEMENT_MODE_INLINE )
-				return;
+		if ( config.fullPage ) {
+			return editor.getData().replace( /<head>/, '$&' + baseTag )
+				.replace( /[^>]*(?=<\/title>)/, '$& &mdash; ' + title );
+		}
 
-			pluginPath = this.path;
+		return config.docType + '<html dir="' + config.contentsLangDirection + '">' +
+			'<head>' +
+				baseTag +
+				'<title>' + title + '</title>' +
+				CKEDITOR.tools.buildStyleHtml( config.contentsCss ) +
+				'<link rel="stylesheet" media="screen" href="' + pluginPath + 'styles/screen.css">' +
+			'</head>' + createBodyHtml() +
+				editor.getData() +
+			'</body></html>';
 
-			editor.addCommand( pluginName, previewCmd );
-			editor.ui.addButton && editor.ui.addButton( 'Preview', {
-				label: editor.lang.preview.preview,
-				command: pluginName,
-				toolbar: 'document,40'
-			} );
+		function generateBaseTag() {
+			var template = '<base href="{HREF}">',
+				origin = location.origin,
+				pathname = location.pathname;
+
+			if ( !config.baseHref && !CKEDITOR.env.gecko ) {
+				return '';
+			}
+
+			if ( config.baseHref ) {
+				return template.replace( '{HREF}', config.baseHref );
+			}
+
+			// As we use HTML file in Gecko, we must fix the incorrect base for
+			// relative paths.
+			pathname = pathname.split( '/' );
+			pathname.pop();
+			pathname = pathname.join( '/' );
+
+			return template.replace( '{HREF}', origin + pathname + '/' );
 		}
-	} );
+
+		function createBodyHtml() {
+			var html = '<body>',
+				body = editor.document && editor.document.getBody();
+
+			if ( !body ) {
+				return html;
+			}
+
+			if ( body.getAttribute( 'id' ) ) {
+				html = html.replace( '>', ' id="' + body.getAttribute( 'id' ) + '">' );
+			}
+
+			if ( body.getAttribute( 'class' ) ) {
+				html = html.replace( '>', ' class="' + body.getAttribute( 'class' ) + '">' );
+			}
+
+			return html;
+		}
+	}
+
+	function getWindowDimensions() {
+		try {
+			var screen = window.screen;
+
+			return {
+				width: Math.round( screen.width * 0.8 ),
+				height: Math.round( screen.height * 0.7 ),
+				left: Math.round( screen.width * 0.1 )
+			};
+		} catch ( e ) {
+			return {
+				width: 640,
+				// 800 * 0.8,
+				height: 420,
+				// 600 * 0.7,
+				left: 80 // (800 - 0.8 * 800) /2 = 800 * 0.1
+			};
+		}
+	}
+
+	function generateWindowOptions( dimensions ) {
+		return [
+			'toolbar=yes',
+			'location=no',
+			'status=yes',
+			'menubar=yes',
+			'scrollbars=yes',
+			'resizable=yes',
+			'width=' + dimensions.width,
+			'height=' + dimensions.height,
+			'left=' + dimensions.left
+		].join( ',' );
+	}
+
+	function getPreviewLocation() {
+		if ( !CKEDITOR.env.ie ) {
+			return;
```

Comment:
As the function returns some value, it would be nice to return `null` here instead of implicit `undefied` for readability.